xmppserver/src/test/java/org/jivesoftware/openfire/pubsub/CachingPubsubPersistenceProviderSubscriptionOperationsTest.java

```diff
@@ -0,0 +1,740 @@
+        final JID subscriber = new JID("test-user-1", "example.org", null);
+        final String subscriptionId = "test-subscription-id-" + StringUtils.randomString(7);
+        final NodeSubscription subscription = new NodeSubscription(mockNode, subscriber, subscriber, NodeSubscription.State.subscribed, subscriptionId);
+
+        // Execute system under test.
+        provider.updateSubscription(mockNode, subscription);
+        provider.removeSubscription(subscription);
+
+        // Verify results.
+        final List<CachingPubsubPersistenceProvider.NodeOperation> pending = provider.nodesToProcess.computeIfAbsent(mockNode.getUniqueIdentifier(), id -> new ArrayList<>());
+        assertEquals(1, pending.size());
+
+        final CachingPubsubPersistenceProvider.NodeOperation nodeOperation = pending.get(0);
+        assertEquals(mockNode.getUniqueIdentifier(), nodeOperation.node.getUniqueIdentifier());
+        assertEquals(CachingPubsubPersistenceProvider.NodeOperation.Action.REMOVE_SUBSCRIPTION, nodeOperation.action);
+        assertEquals(subscriptionId, nodeOperation.subscription.getID());
+        assertNull(nodeOperation.affiliate);
+
+        assertEquals(0, provider.itemsPending.size());
+        assertEquals(0, provider.itemsToAdd.size());
+        assertEquals(0, provider.itemsToDelete.size());
+    }
+
+
+    /**
+     * Executes a test that, through {@link CachingPubsubPersistenceProvider#createSubscription(Node, NodeSubscription)},
+     * {@link CachingPubsubPersistenceProvider#updateSubscription(Node, NodeSubscription)} and
+     * {@link CachingPubsubPersistenceProvider#removeSubscription(NodeSubscription)} a subscription is created, updated
+     * and immediately removes again.
+     *
+     * The caching provider can optimize these three operations, where the 'net effect' of them is to have no operation.
+     *
+     * Therefor, this test asserts that after each invocation, no operation is scheduled.
+     */
+    @Test
+    public void testDeleteSubscriptionVoidsOlderCreateWithUpdate() throws Exception
+    {
+        // Setup test fixture.
+        final CachingPubsubPersistenceProvider provider = new CachingPubsubPersistenceProvider();
+        final Node mockNode = Mockito.mock(Node.class);
+        Mockito.lenient().when(mockNode.getUniqueIdentifier()).thenReturn(new Node.UniqueIdentifier("mock-service-1", "mock-node-a"));
+        final JID subscriber = new JID("test-user-1", "example.org", null);
+        final String subscriptionId = "test-subscription-id-" + StringUtils.randomString(7);
+        final NodeSubscription subscription = new NodeSubscription(mockNode, subscriber, subscriber, NodeSubscription.State.subscribed, subscriptionId);
+
+        // Execute system under test.
+        provider.createSubscription(mockNode, subscription);
+        provider.updateSubscription(mockNode, subscription);
+        provider.removeSubscription(subscription);
```

Comment:
What if the create is rejected because the subscription already exists?
The update could be optimised away, but the remove might still be needed, right?

Although... Can an update ever occur that prevents removing the subscription? Like updating an owner?

Reply:
I've been debating things like this while I was coding this. I tried to make decisions about a very basic aspect of the system-under-test: it is strictly a data-access-object, in the sense that its sole responsibility is to interact with persistent storage. In that sense, it has less responsibility to verify that the invocation is correct (eg: flag that a creation should fail because it already exists). I think what you describe is something that's on the edge of that balance. I'm not really convinced of either approach to be honest.